test/gtest/conv_immediate.cpp

```diff
@@ -0,0 +1,532 @@
+                            int groupCount,
+                            miopenConvolutionMode_t mode = miopenConvolution)
+{
+    miopenConvolutionDescriptor_t desc;
+
+    EXPECT_TRUE(miopenCreateConvolutionDescriptor(&desc) == miopenStatusSuccess);
+    EXPECT_TRUE(miopenInitConvolutionNdDescriptor(
+                    desc, pads.size(), pads.data(), strides.data(), dilations.data(), mode) ==
+                miopenStatusSuccess);
+    EXPECT_TRUE(miopenSetConvolutionGroupCount(desc, groupCount) == miopenStatusSuccess);
+
+    return desc;
+}
+
+template <miopen::conv::Direction direction,
+          typename TREF,
+          typename TOUT,
+          miopenTensorLayout_t tensor_layout>
+struct TypeDefs
+{
+    static constexpr miopen::conv::Direction Direction = direction;
+    using TRef                                         = TREF;
+    using TOut                                         = TOUT;
+    static constexpr miopenTensorLayout_t Layout       = tensor_layout;
+};
+
+template <typename T>
+struct layout_data
+{
+    static std::vector<int> get_layout_lengths(int n, int c, std::vector<int>& dims)
+    {
+        auto ret = std::vector<int>{n, c};
+        ret.insert(ret.end(), dims.cbegin(), dims.cend());
+
+        return ret;
+    }
+
+    static std::vector<int>
+    get_strides(std::vector<int>& lens, int dims, miopenTensorLayout_t tensor_layout)
+    {
+        std::vector<int> strides;
+        std::string layout_default = miopen::tensor_layout_get_default(dims + 2);
+        std::string layout_string  = miopen::TensorDescriptor::GetLayoutStr(tensor_layout);
+
+        miopen::tensor_layout_to_strides(lens, layout_default, layout_string, strides);
+
+        constexpr int min_stride_multiplier = 1;
+        constexpr int max_stride_multiplier = 4;
+
+        auto c = prng::gen_A_to_B(min_stride_multiplier, max_stride_multiplier);
+        for(auto& v : strides)
+        {
+            // cppcheck-suppress useStlAlgorithm
+            v = v * c;
+        }
+
+        return strides;
+    }
+
+    static miopenTensorDescriptor_t init_tensor_descriptor(miopenDataType_t type,
+                                                           const std::vector<int>& lens,
+                                                           const std::vector<int>& strides)
+    {
+        miopenTensorDescriptor_t desc;
+
+        EXPECT_TRUE(miopenCreateTensorDescriptor(&desc) == miopenStatusSuccess);
+        EXPECT_TRUE(
+            miopenSetTensorDescriptor(desc, type, lens.size(), lens.data(), strides.data()) ==
+            miopenStatusSuccess);
+
+        return desc;
+    }
+
+    layout_data(int _n, std::vector<int> _dims, int _c, miopenTensorLayout_t _tensor_layout)
+    {
+        auto lens    = get_layout_lengths(_n, _c, _dims);
+        auto strides = get_strides(lens, _dims.size(), _tensor_layout);
+        descriptor   = miopen::TensorDescriptor{miopen_type<T>{}, lens, strides};
```

Comment:
as @CAHEK7  suggested, it's better to use `tensor<T>` than `TensorDescriptor`.

Reply:
Yes, please refer to my reply below.

Reply:
On the 'two copies of B' question: I was just replicating the original test cases which use repeated values for h and w for image, filter, and pad sizes. It ends up running two squares and two rectangles of each. For example, image sizes tested are [9x9], [14x9], [9x14], and [14x14].